s3transfer/futures.py

```diff
@@ -422,38 +536,91 @@ def shutdown(self, wait=True):
-        However, this can eventually grow to implement the needed functionality
-        of concurrent.futures.Future if we move off of the library and not
-        affect the rest of the codebase.
-
-        :type future: concurrent.futures.Future
-        :param future: The underlying future
+        This is a pared down version of concurrent.futures.Future, stripping
+        away all the features that we don't use and modifying the interface
+        somewhat (namely removing the argument from callbacks).
         """
-        self._future = future
+        self._condition = threading.Condition()
+        self._done = False
+        self._result = None
+        self._exception = None
+        self._traceback = None
+        self._done_callbacks = []
 
-    def result(self):
-        return self._future.result()
+    def _invoke_callbacks(self):
+        """Invokes all the queued done callbacks."""
+        for callback in self._done_callbacks:
+            try:
+                callback()
+            except Exception:
+                logger.debug('exception calling callback for %r', self)
+
+    def done(self):
+        """Return True if the future has finished."""
+        with self._condition:
+            return self._done
+
+    def _get_result(self):
+        """Returns the future's result or reraises the thrown exception."""
+        if self._exception:
+            six.reraise(
+                type(self._exception), self._exception, self._traceback)
+        else:
+            return self._result
 
     def add_done_callback(self, fn):
-        """Adds a callback to be completed once future is done
+        """Add a callback to be called when the future is done.
 
-        :parm fn: A callable that takes no arguments. Note that is different
+        :type fn: function
+        :param fn: A callable that takes no arguments. Note that is different
             than concurrent.futures.Future.add_done_callback that requires
             a single argument for the future.
         """
-        # The done callback for concurrent.futures.Future will always pass a
-        # the future in as the only argument. So we need to create the
-        # proper signature wrapper that will invoke the callback provided.
-        def done_callback(future_passed_to_callback):
-            return fn()
-        self._future.add_done_callback(done_callback)
+        with self._condition:
+            if not self._done:
+                self._done_callbacks.append(fn)
+                return
+        # Call the callback now if the future is already done.
+        fn()
+
+    def result(self, timeout=None):
+        """Return the result of the future.
+
+        :type timeout: int
+        :param timeout: A number of seconds to wait for the result. A
+            TimeoutError will be raised if the future is not done in time.
+        """
+        with self._condition:
+            if self._done:
```

Comment:
Do yo know why this is done twice: once before the wait and one after the wait:
```python
if self._done:
     return self._get_result()
```
I feel like it will work with that statement after the wait.